File: ios/Classes/PlatformComparisonPlugin.swift
import Flutter

public class PlatformComparisonPlugin: NSObject, FlutterPlugin {

    private var channel: FlutterMethodChannel? = nil

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "platform_comparison_plugin", 
            binaryMessenger: registrar.messenger())
        let instance = PlatformComparisonPlugin()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: instance.channel!)
        registrar.addApplicationDelegate(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if (call.method == "start") {

            let arguments = call.arguments as! [String: Any]
            let id = arguments["id"] as? String
            
            self.channel!.invokeMethod("processing", arguments: nil, result: {(r:Any?) -> () in })

            validateIDForResult(id: id)
        } else {
            result(FlutterMethodNotImplemented)
        }
    }

    private func validateIDForResult(id: String?) -> Void {
        // let's mock a delay of 3 seconds on platform side   
        DispatchQueue.main.asyncAfter(deadline: .now() + 3) { 
            //simulate a validation process on a platform side
            if (id != nil && id == "IOS") {
                // optionally, if we were into the simplest interop approach,
                // we could pass a result back to the Flutter side 
                // as Flutter iOS Embedder's FlutterResult supports dual use:
                //result("iOS: start() method called")
                
                self.channel!.invokeMethod("success", arguments: nil, result: {(r:Any?) -> () in })
            } else {
                self.channel!.invokeMethod("error", arguments: nil, result: {(r:Any?) -> () in })
            }
        }
    }
}

File: android/src/main/kotlin/com/example/platform_comparison_plugin/PlatformComparisonPlugin.kt
package com.example.platform_comparison_plugin

import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import android.os.Handler
import android.os.Looper

class PlatformComparisonPlugin: FlutterPlugin, MethodCallHandler {

  private lateinit var channel : MethodChannel

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
      channel = MethodChannel(flutterPluginBinding.binaryMessenger, "platform_comparison_plugin")
      channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
    when (call.method) {
        "start" -> {

          val id: String? = call.argument("id")

          channel.invokeMethod("processing", null)

          validateIDForResult(id)
          
        } else -> {
          result.notImplemented()
        }
    }
  }

  private fun validateIDForResult(id: String?) {
    //let's mock a delay of 3 seconds on platform side    
    Handler(Looper.getMainLooper()).postDelayed({
        //simulate a validation process on a platform side
        if (id != null && id == "ANDROID") {
            //optionally, if we were into the simplest interop approach,
            // we could pass a result back to the Flutter side 
            // as MethodChannel's result supports dual use:
            //result.success("Android - start() method called")

            channel.invokeMethod("success", null)
        } else {
            channel.invokeMethod("error", null)
        }
    }, 3000)
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

}
